java/code/src/com/suse/manager/webui/controllers/contentmanagement/handlers/ProjectSourcesApiController.java

```diff
@@ -0,0 +1,110 @@
+ * version 2 (GPLv2). There is NO WARRANTY for this software, express or
+ * implied, including the implied warranties of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+ * along with this software; if not, see
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+ *
+ * Red Hat trademarks are not licensed under GPLv2. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package com.suse.manager.webui.controllers.contentmanagement.handlers;
+
+import static com.suse.manager.webui.utils.SparkApplicationHelper.json;
+import static com.suse.manager.webui.utils.SparkApplicationHelper.withUser;
+import static spark.Spark.put;
+
+import com.redhat.rhn.domain.contentmgmt.ContentEnvironment;
+import com.redhat.rhn.domain.contentmgmt.ContentProject;
+import com.redhat.rhn.domain.contentmgmt.ProjectSource;
+import com.redhat.rhn.domain.contentmgmt.SoftwareProjectSource;
+import com.redhat.rhn.domain.user.User;
+import com.redhat.rhn.manager.contentmgmt.ContentManager;
+
+import com.suse.manager.webui.controllers.contentmanagement.mappers.ResponseMappers;
+import com.suse.manager.webui.controllers.contentmanagement.request.ProjectSourcesRequest;
+import com.suse.manager.webui.utils.gson.ResultJson;
+import com.suse.utils.Json;
+
+import com.google.gson.Gson;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import spark.Request;
+import spark.Response;
+
+/**
+ * Spark controller class for content management sources API endpoints.
+ */
+public class ProjectSourcesApiController {
+
+    private static final Gson GSON = Json.GSON;
+
+    private ProjectSourcesApiController() {
+    }
+
+    /** Init routes for ContentManagement Sources Api.*/
+    public static void initRoutes() {
+        put("/manager/contentmanagement/api/projects/:projectId/softwaresources",
+                withUser(ProjectSourcesApiController::updateContentSoftwareSources));
+
+    }
+
+    /**
+     * Return the JSON with the result of updating the content project with a list of sources.
+     * @param req the http request
+     * @param res the http response
+     * @param user the current user
+     * @return the JSON data
+     */
+    public static String updateContentSoftwareSources(Request req, Response res, User user) {
+        ProjectSourcesRequest createSourceRequest = ProjectSourcesHandler.getSourcesRequest(req);
+        String projectLabel = createSourceRequest.getProjectLabel();
+
+        ContentProject dbContentProject = ContentManager.lookupProject(
+                createSourceRequest.getProjectLabel(), user
+        ).get();
+
+        List<String> sourceLabelsToDetach = dbContentProject.getSources()
+                .stream()
+                .filter(SoftwareProjectSource.class::isInstance)
+                .map(SoftwareProjectSource.class::cast)
```

Comment:
You could use `Source::asSoftwareSource` for filtering, something like in the `ContentManager`:

```java
// something like this
project.getSources().stream()
            .flatMap(s -> stream(s.asSoftwareSource()))
            .map(s -> s.getChannel().getLabel());
```